tests/aio.rs

```diff
@@ -0,0 +1,226 @@
+        // number and name of reserved fields is OS-dependent.  On some OSes,
+        // some reserved fields are used the kernel for state, and must be
+        // explicitly zeroed when allocated.
+        let mut inner = unsafe { mem::zeroed::<libc::aiocb>() };
+        inner.aio_fildes = fd;
+        inner.aio_sigevent.sigev_notify = libc::SIGEV_NONE;
+        inner.aio_offset = 0;
+        inner.aio_nbytes = 0;
+        Aiocb(Box::pin(inner))
+    }
+
+    /// Constructs a new `Aiocb` suitable for writing to offset 0 of a file.
+    pub fn from_slice(fd: RawFd, buf: &[u8]) -> Aiocb {
+        let mut aiocb = Aiocb::from_fd(fd);
+        aiocb.0.aio_nbytes = buf.len();
+        aiocb.0.aio_buf = buf.as_ptr() as *mut libc::c_void;
+        aiocb
+    }
+
+    pub fn fsync(&mut self) -> Result<(), i32> {
+        unsafe {
+            // Safe because we don't move the libc::aiocb
+            let selfp = self.0.as_mut().get_unchecked_mut();
+            match libc::aio_fsync(libc::O_SYNC, selfp) {
+                0 => Ok(()),
+                _ => Err(*libc::__error()),
+            }
+        }
+    }
+}
+
+impl Source for Aiocb {
+    fn register(
+        &mut self,
+        registry: &Registry,
+        token: Token,
+        interests: Interest,
+    ) -> io::Result<()> {
+        assert!(interests.is_aio());
+        let udata = usize::from(token);
+        let kq = registry.as_raw_fd();
+        self.0.aio_sigevent.sigev_notify = libc::SIGEV_KEVENT;
+        self.0.aio_sigevent.sigev_signo = kq;
+        self.0.aio_sigevent.sigev_value.sival_ptr = udata as *mut libc::c_void;
+        self.0.aio_sigevent.sigev_notify_thread_id = 0;
+        Ok(())
+    }
+
+    fn reregister(
+        &mut self,
+        registry: &Registry,
+        token: Token,
+        interests: Interest,
+    ) -> io::Result<()> {
+        self.register(registry, token, interests)
+    }
+
+    fn deregister(&mut self, _registry: &Registry) -> io::Result<()> {
+        self.0.aio_sigevent.sigev_notify = libc::SIGEV_NONE;
+        self.0.aio_sigevent.sigev_value.sival_ptr = ptr::null_mut();
+        Ok(())
+    }
+}
+
+#[cfg(target_os = "freebsd")]
+struct Liocb {
+    _aiocbs: Box<[Aiocb]>,
+    /// The actual list passed to `libc::lio_listio`.
+    ///
+    /// It must live for as long as any of the operations are still being
+    /// processesed, because the aio subsystem uses its address as a unique
+    /// identifier.
+    list: Box<[*mut libc::aiocb]>,
+    sev: libc::sigevent,
+}
+
+#[cfg(target_os = "freebsd")]
+impl Liocb {
+    fn listio(&mut self) -> Result<(), i32> {
+        unsafe {
+            let r = libc::lio_listio(
+                libc::LIO_NOWAIT,
+                self.list.as_ptr(),
+                self.list.len() as i32,
+                &mut self.sev as *mut libc::sigevent,
+            );
+            match r {
+                0 => Ok(()),
+                _ => Err(*libc::__error()),
```

Comment:
Nit: same as above, why not use `io::Error::last_os_error` and return an `io::Error`?